tools/engine_tool/lib/src/typed_json.dart

```diff
@@ -0,0 +1,346 @@
+      _error(InvalidTypeJsonReadException(
+        this,
+        key,
+        expected: T,
+        actual: value.runtimeType,
+      ));
+      return null;
+    } else {
+      return value;
+    }
+  }
+
+  /// Returns the value at the given [key] as a [String].
+  ///
+  /// If the value is not found, returns `null`.
+  ///
+  /// Throws a [JsonReadException] if the value cannot be cast.
+  String? stringOrNull(String key) => _getOrNull<String>(key);
+
+  /// Returns the value at the given [key] as an [int].
+  ///
+  /// If the value is not found, returns `null`.
+  ///
+  /// Throws a [JsonReadException] if the value cannot be cast.
+  int? integerOrNull(String key) => _getOrNull<int>(key);
+
+  /// Returns the value at the given [key] as a [bool].
+  ///
+  /// If the value is not found, returns `null`.
+  ///
+  /// Throws a [JsonReadException] if the value cannot be cast.
+  bool? booleanOrNull(String key) => _getOrNull<bool>(key);
+
+  List<T>? _getListOrNull<T extends Object>(String key) {
+    final Object? value = _object[key];
+    if (value == null && !_object.containsKey(key)) {
+      return null;
+    } else if (value is! List<Object?>) {
+      _error(InvalidTypeJsonReadException(
+        this,
+        key,
+        expected: List<T>,
+        actual: value.runtimeType,
+      ));
+      return <T>[];
+    } else {
+      final List<T> result = <T>[];
+      for (final Object? element in value) {
+        if (element is! T) {
+          _error(InvalidTypeJsonReadException(
+            this,
+            key,
+            expected: T,
+            actual: element.runtimeType,
+          ));
+          return <T>[];
+        }
+        result.add(element);
+      }
+      return result;
+    }
+  }
+
+  /// Returns the value at the given [key] as a [List] of [String]s.
+  ///
+  /// If the value is not found, returns `null`.
+  ///
+  /// Throws a [JsonReadException] if the value cannot be cast.
+  List<String>? stringListOrNull(String key) => _getListOrNull<String>(key);
+
+  static Never _onErrorThrow(JsonObject object, JsonMapException e) {
+    throw e;
+  }
+
+  /// Returns the result of applying the given [mapper] to this JSON object.
+  ///
+  /// Any exception that otherwise would be thrown by the mapper is caught
+  /// internally and rethrown as a [JsonMapException] with all of the original
+  /// exceptions added to its [exceptions] property.
+  ///
+  /// If [onError] is provided, it is called with this object and the caught
+  /// [JsonMapException] to handle the error. If not provided, the exception is
+  /// rethrown.
+  ///
+  /// ## Example
+  ///
+  /// ```dart
+  /// final person = json.map(
+  ///   (json) => Person(
+  ///     name: json.string('name'),
+  ///     age: json.integer('age'),
+  ///     isEmployed: json.boolean('isEmployed'),
+  ///   ), onError: (json, e) {
+  ///     throw FormatException('Failed to read person: $e');
+  ///   },
+  /// );
+  /// ```
+  T map<T extends Object>(
```

Comment:
I'm not sure how it would be useful, but if `T` can be a `Future`, then I'm not sure that the mechanism for saving and restoring the errors will work. Is there some way to say that `T` can't be a `Future`?

Reply:
2018 Matan was ahead of us both: https://github.com/dart-lang/sdk/issues/35024.

It is probably trivial to add such a hint assuming the analyzer team is on board. For now I'll do a runtime check that the result is not a Future.